<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Survival Analysis for Colorectal Cancer Patients</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="Colorectal-Cancer-Survival-Analysis_files/libs/clipboard/clipboard.min.js"></script>
<script src="Colorectal-Cancer-Survival-Analysis_files/libs/quarto-html/quarto.js"></script>
<script src="Colorectal-Cancer-Survival-Analysis_files/libs/quarto-html/popper.min.js"></script>
<script src="Colorectal-Cancer-Survival-Analysis_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Colorectal-Cancer-Survival-Analysis_files/libs/quarto-html/anchor.min.js"></script>
<link href="Colorectal-Cancer-Survival-Analysis_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Colorectal-Cancer-Survival-Analysis_files/libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Colorectal-Cancer-Survival-Analysis_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Colorectal-Cancer-Survival-Analysis_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Colorectal-Cancer-Survival-Analysis_files/libs/bootstrap/bootstrap-973236bd072d72a04ee9cd82dcc9cb29.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><strong>Survival Analysis for Colorectal Cancer Patients</strong></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="imports" class="level2">
<h2 class="anchored" data-anchor-id="imports"><strong>Imports</strong></h2>
<div id="0e26b58d-7bd6-4a6a-94a2-caadf04c9c7e" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.decomposition <span class="im">import</span> PCA</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.compose <span class="im">import</span> ColumnTransformer, make_column_transformer</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.dummy <span class="im">import</span> DummyClassifier</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.ensemble <span class="im">import</span> RandomForestClassifier, RandomForestRegressor</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.impute <span class="im">import</span> SimpleImputer</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.linear_model <span class="im">import</span> LogisticRegression, Ridge</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.metrics <span class="im">import</span> confusion_matrix</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.model_selection <span class="im">import</span> (</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>    cross_val_predict,</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>    cross_val_score,</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>    cross_validate,</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>    train_test_split,</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.pipeline <span class="im">import</span> Pipeline, make_pipeline</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.preprocessing <span class="im">import</span> (</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>    FunctionTransformer,</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>    OneHotEncoder,</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>    OrdinalEncoder,</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>    StandardScaler,</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> lifelines</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="eda" class="level2">
<h2 class="anchored" data-anchor-id="eda"><strong>EDA</strong></h2>
<div id="1337b2ba-5668-4401-a335-fd997e66ff09" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>patient_data <span class="op">=</span> pd.read_csv(<span class="st">"data/Colorectal Cancer Patient Data.csv"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>gene_expression_data <span class="op">=</span> pd.read_csv(<span class="st">"data/Colorectal Cancer Gene Expression Data.csv"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>display(patient_data.info())</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>display(gene_expression_data.info())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;class 'pandas.core.frame.DataFrame'&gt;
RangeIndex: 63 entries, 0 to 62
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       63 non-null     int64  
 1   ID_REF           62 non-null     object 
 2   Age (in years)   62 non-null     float64
 3   Dukes Stage      62 non-null     object 
 4   Gender           62 non-null     object 
 5   Location         62 non-null     object 
 6   DFS (in months)  62 non-null     float64
 7   DFS event        62 non-null     float64
 8   Adj_Radio        62 non-null     float64
 9   Adj_Chem         62 non-null     float64
dtypes: float64(5), int64(1), object(4)
memory usage: 5.1+ KB</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>None</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;class 'pandas.core.frame.DataFrame'&gt;
RangeIndex: 1935 entries, 0 to 1934
Data columns (total 64 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  1935 non-null   int64  
 1   ID_REF      1935 non-null   object 
 2   GSM877126   1935 non-null   float64
 3   GSM877127   1935 non-null   float64
 4   GSM877128   1935 non-null   float64
 5   GSM877129   1935 non-null   float64
 6   GSM877130   1935 non-null   float64
 7   GSM877131   1935 non-null   float64
 8   GSM877132   1935 non-null   float64
 9   GSM877133   1935 non-null   float64
 10  GSM877134   1935 non-null   float64
 11  GSM877135   1935 non-null   float64
 12  GSM877136   1935 non-null   float64
 13  GSM877137   1935 non-null   float64
 14  GSM877138   1935 non-null   float64
 15  GSM877139   1935 non-null   float64
 16  GSM877140   1935 non-null   float64
 17  GSM877141   1935 non-null   float64
 18  GSM877142   1935 non-null   float64
 19  GSM877143   1935 non-null   float64
 20  GSM877144   1935 non-null   float64
 21  GSM877145   1935 non-null   float64
 22  GSM877146   1935 non-null   float64
 23  GSM877147   1935 non-null   float64
 24  GSM877148   1935 non-null   float64
 25  GSM877149   1935 non-null   float64
 26  GSM877150   1935 non-null   float64
 27  GSM877151   1935 non-null   float64
 28  GSM877152   1935 non-null   float64
 29  GSM877153   1935 non-null   float64
 30  GSM877154   1935 non-null   float64
 31  GSM877155   1935 non-null   float64
 32  GSM877156   1935 non-null   float64
 33  GSM877157   1935 non-null   float64
 34  GSM877158   1935 non-null   float64
 35  GSM877159   1935 non-null   float64
 36  GSM877160   1935 non-null   float64
 37  GSM877161   1935 non-null   float64
 38  GSM877162   1935 non-null   float64
 39  GSM877163   1935 non-null   float64
 40  GSM877164   1935 non-null   float64
 41  GSM877165   1935 non-null   float64
 42  GSM877166   1935 non-null   float64
 43  GSM877167   1935 non-null   float64
 44  GSM877168   1935 non-null   float64
 45  GSM877169   1935 non-null   float64
 46  GSM877170   1935 non-null   float64
 47  GSM877171   1935 non-null   float64
 48  GSM877173   1935 non-null   float64
 49  GSM877174   1935 non-null   float64
 50  GSM877175   1935 non-null   float64
 51  GSM877176   1935 non-null   float64
 52  GSM877177   1935 non-null   float64
 53  GSM877178   1935 non-null   float64
 54  GSM877179   1935 non-null   float64
 55  GSM877180   1935 non-null   float64
 56  GSM877181   1935 non-null   float64
 57  GSM877182   1935 non-null   float64
 58  GSM877183   1935 non-null   float64
 59  GSM877184   1935 non-null   float64
 60  GSM877185   1935 non-null   float64
 61  GSM877186   1935 non-null   float64
 62  GSM877187   1935 non-null   float64
 63  GSM877188   1935 non-null   float64
dtypes: float64(62), int64(1), object(1)
memory usage: 967.6+ KB</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>None</code></pre>
</div>
</div>
<div id="a6ccc78d-6f33-4c3b-8f4f-8827421c63e4" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Catch ID_REF from patient_data</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>patient_ids <span class="op">=</span> <span class="bu">list</span>(patient_data[<span class="st">'ID_REF'</span>].dropna())</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Select relative columns from gene_expression_data</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>filtered_gene_expression <span class="op">=</span> gene_expression_data[[<span class="st">'ID_REF'</span>] <span class="op">+</span> patient_ids]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="c910725c-e0a1-45e6-9aae-a08c9420bdcf" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Transposed gene_expression_data</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>transposed_gene_expression <span class="op">=</span> filtered_gene_expression.set_index(<span class="st">'ID_REF'</span>).T</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Use PCA to decreasing the dimension</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>pca <span class="op">=</span> PCA(n_components<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>gene_data_pca <span class="op">=</span> pca.fit_transform(transposed_gene_expression)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>gene_data_pca_df <span class="op">=</span> pd.DataFrame(</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>    gene_data_pca, </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>    index<span class="op">=</span>transposed_gene_expression.index,</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>    columns<span class="op">=</span>[<span class="ss">f'PC</span><span class="sc">{</span>i<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">'</span> <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(gene_data_pca.shape[<span class="dv">1</span>])]</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>gene_data_pca_df.index.name <span class="op">=</span> <span class="st">'ID_REF'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="f3944a9e-cc6b-49b6-80e3-33929d057044" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge data by ID_REF</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>cancer_df <span class="op">=</span> pd.merge(patient_data, gene_data_pca_df, on<span class="op">=</span><span class="st">'ID_REF'</span>, how<span class="op">=</span><span class="st">'inner'</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>display(cancer_df.head())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Unnamed: 0</th>
<th data-quarto-table-cell-role="th">ID_REF</th>
<th data-quarto-table-cell-role="th">Age (in years)</th>
<th data-quarto-table-cell-role="th">Dukes Stage</th>
<th data-quarto-table-cell-role="th">Gender</th>
<th data-quarto-table-cell-role="th">Location</th>
<th data-quarto-table-cell-role="th">DFS (in months)</th>
<th data-quarto-table-cell-role="th">DFS event</th>
<th data-quarto-table-cell-role="th">Adj_Radio</th>
<th data-quarto-table-cell-role="th">Adj_Chem</th>
<th data-quarto-table-cell-role="th">PC1</th>
<th data-quarto-table-cell-role="th">PC2</th>
<th data-quarto-table-cell-role="th">PC3</th>
<th data-quarto-table-cell-role="th">PC4</th>
<th data-quarto-table-cell-role="th">PC5</th>
<th data-quarto-table-cell-role="th">PC6</th>
<th data-quarto-table-cell-role="th">PC7</th>
<th data-quarto-table-cell-role="th">PC8</th>
<th data-quarto-table-cell-role="th">PC9</th>
<th data-quarto-table-cell-role="th">PC10</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>0</td>
<td>GSM877126</td>
<td>62.0</td>
<td>A</td>
<td>Male</td>
<td>Left</td>
<td>108.0</td>
<td>0.0</td>
<td>1.0</td>
<td>0.0</td>
<td>10.034928</td>
<td>-2.563461</td>
<td>0.553594</td>
<td>3.786064</td>
<td>5.491033</td>
<td>2.113068</td>
<td>0.850123</td>
<td>-1.365781</td>
<td>-1.001487</td>
<td>0.467374</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>1</td>
<td>GSM877127</td>
<td>77.0</td>
<td>B</td>
<td>Male</td>
<td>Left</td>
<td>40.0</td>
<td>1.0</td>
<td>1.0</td>
<td>0.0</td>
<td>3.952900</td>
<td>5.896938</td>
<td>-9.498808</td>
<td>-8.151790</td>
<td>-1.408558</td>
<td>3.904793</td>
<td>2.074637</td>
<td>3.621052</td>
<td>1.138539</td>
<td>4.319512</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>2</td>
<td>GSM877128</td>
<td>66.0</td>
<td>C</td>
<td>Female</td>
<td>Left</td>
<td>49.0</td>
<td>0.0</td>
<td>1.0</td>
<td>0.0</td>
<td>5.775725</td>
<td>-3.307617</td>
<td>-1.514136</td>
<td>1.656845</td>
<td>0.277657</td>
<td>0.466875</td>
<td>-0.887899</td>
<td>2.521145</td>
<td>-6.819103</td>
<td>2.272644</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>3</td>
<td>GSM877129</td>
<td>72.0</td>
<td>D</td>
<td>Female</td>
<td>Left</td>
<td>45.0</td>
<td>0.0</td>
<td>1.0</td>
<td>1.0</td>
<td>12.069910</td>
<td>-3.575149</td>
<td>0.164506</td>
<td>0.873292</td>
<td>-0.304559</td>
<td>-1.279556</td>
<td>-1.271283</td>
<td>4.878356</td>
<td>0.760081</td>
<td>-4.307443</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>4</td>
<td>GSM877130</td>
<td>75.0</td>
<td>C</td>
<td>Male</td>
<td>Left</td>
<td>40.0</td>
<td>0.0</td>
<td>0.0</td>
<td>1.0</td>
<td>0.704046</td>
<td>5.934426</td>
<td>4.480248</td>
<td>-1.633834</td>
<td>2.685619</td>
<td>0.053008</td>
<td>3.949680</td>
<td>2.330419</td>
<td>0.315755</td>
<td>0.129697</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="c0bb31b4-d698-43a5-a323-193763cf290d" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Drop columns: "unnamed: 0", "ID_REF"</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>cancer_df <span class="op">=</span> cancer_df.drop(columns<span class="op">=</span>[<span class="st">"Unnamed: 0"</span>, <span class="st">"ID_REF"</span>], errors<span class="op">=</span><span class="st">'ignore'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="4bc93dbd-603d-420b-9abb-2647f8f9cfeb" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>display(cancer_df.info())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;class 'pandas.core.frame.DataFrame'&gt;
RangeIndex: 62 entries, 0 to 61
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Age (in years)   62 non-null     float64
 1   Dukes Stage      62 non-null     object 
 2   Gender           62 non-null     object 
 3   Location         62 non-null     object 
 4   DFS (in months)  62 non-null     float64
 5   DFS event        62 non-null     float64
 6   Adj_Radio        62 non-null     float64
 7   Adj_Chem         62 non-null     float64
 8   PC1              62 non-null     float64
 9   PC2              62 non-null     float64
 10  PC3              62 non-null     float64
 11  PC4              62 non-null     float64
 12  PC5              62 non-null     float64
 13  PC6              62 non-null     float64
 14  PC7              62 non-null     float64
 15  PC8              62 non-null     float64
 16  PC9              62 non-null     float64
 17  PC10             62 non-null     float64
dtypes: float64(15), object(3)
memory usage: 8.8+ KB</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>None</code></pre>
</div>
</div>
<div id="b4248dfd-4cf6-4d69-b7fb-676e4c83f429" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>display(cancer_df.describe())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Age (in years)</th>
<th data-quarto-table-cell-role="th">DFS (in months)</th>
<th data-quarto-table-cell-role="th">DFS event</th>
<th data-quarto-table-cell-role="th">Adj_Radio</th>
<th data-quarto-table-cell-role="th">Adj_Chem</th>
<th data-quarto-table-cell-role="th">PC1</th>
<th data-quarto-table-cell-role="th">PC2</th>
<th data-quarto-table-cell-role="th">PC3</th>
<th data-quarto-table-cell-role="th">PC4</th>
<th data-quarto-table-cell-role="th">PC5</th>
<th data-quarto-table-cell-role="th">PC6</th>
<th data-quarto-table-cell-role="th">PC7</th>
<th data-quarto-table-cell-role="th">PC8</th>
<th data-quarto-table-cell-role="th">PC9</th>
<th data-quarto-table-cell-role="th">PC10</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">count</td>
<td>62.000000</td>
<td>62.000000</td>
<td>62.000000</td>
<td>62.000000</td>
<td>62.000000</td>
<td>6.200000e+01</td>
<td>6.200000e+01</td>
<td>6.200000e+01</td>
<td>6.200000e+01</td>
<td>6.200000e+01</td>
<td>6.200000e+01</td>
<td>6.200000e+01</td>
<td>6.200000e+01</td>
<td>6.200000e+01</td>
<td>6.200000e+01</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">mean</td>
<td>61.112903</td>
<td>41.774194</td>
<td>0.596774</td>
<td>0.532258</td>
<td>0.483871</td>
<td>-6.088320e-16</td>
<td>5.730183e-16</td>
<td>1.189013e-15</td>
<td>-8.595275e-16</td>
<td>-5.479488e-16</td>
<td>-9.848753e-16</td>
<td>1.969751e-15</td>
<td>5.318326e-16</td>
<td>1.633102e-15</td>
<td>6.589711e-16</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">std</td>
<td>9.581582</td>
<td>26.288076</td>
<td>0.494550</td>
<td>0.503032</td>
<td>0.503819</td>
<td>5.984986e+00</td>
<td>5.448934e+00</td>
<td>5.289646e+00</td>
<td>3.707437e+00</td>
<td>3.560856e+00</td>
<td>3.043141e+00</td>
<td>2.942446e+00</td>
<td>2.866659e+00</td>
<td>2.747006e+00</td>
<td>2.688011e+00</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">min</td>
<td>28.000000</td>
<td>4.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>-8.734801e+00</td>
<td>-1.009633e+01</td>
<td>-1.348958e+01</td>
<td>-8.151790e+00</td>
<td>-7.374372e+00</td>
<td>-6.927775e+00</td>
<td>-9.725663e+00</td>
<td>-8.668528e+00</td>
<td>-6.819103e+00</td>
<td>-6.508976e+00</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">25%</td>
<td>56.250000</td>
<td>19.500000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>-4.910225e+00</td>
<td>-3.489797e+00</td>
<td>-2.602366e+00</td>
<td>-2.750916e+00</td>
<td>-2.110189e+00</td>
<td>-1.456760e+00</td>
<td>-1.386963e+00</td>
<td>-1.694862e+00</td>
<td>-1.549649e+00</td>
<td>-1.149974e+00</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">50%</td>
<td>62.000000</td>
<td>38.000000</td>
<td>1.000000</td>
<td>1.000000</td>
<td>0.000000</td>
<td>-1.208833e+00</td>
<td>-8.254880e-01</td>
<td>1.111445e-01</td>
<td>-3.605246e-01</td>
<td>-1.906082e-01</td>
<td>5.593847e-02</td>
<td>3.260334e-01</td>
<td>-8.100692e-02</td>
<td>4.634558e-01</td>
<td>2.190788e-01</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">75%</td>
<td>67.000000</td>
<td>57.500000</td>
<td>1.000000</td>
<td>1.000000</td>
<td>1.000000</td>
<td>4.018212e+00</td>
<td>2.732425e+00</td>
<td>2.627977e+00</td>
<td>2.183317e+00</td>
<td>1.288891e+00</td>
<td>2.091801e+00</td>
<td>1.633154e+00</td>
<td>2.266200e+00</td>
<td>1.697653e+00</td>
<td>1.067278e+00</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">max</td>
<td>78.000000</td>
<td>108.000000</td>
<td>1.000000</td>
<td>1.000000</td>
<td>1.000000</td>
<td>1.345370e+01</td>
<td>1.610472e+01</td>
<td>1.591209e+01</td>
<td>9.773375e+00</td>
<td>1.891697e+01</td>
<td>7.146467e+00</td>
<td>6.469778e+00</td>
<td>5.776316e+00</td>
<td>4.685594e+00</td>
<td>8.002474e+00</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<ul>
<li><strong>Age Distribution</strong>:
<ul>
<li>Patients are mostly older (median: 62 years), with a moderately small spread.</li>
</ul></li>
<li><strong>DFS (Disease-Free Survival)</strong>:
<ul>
<li>DFS shows high variability, ranging from very short (4 months) to very long (108 months).</li>
<li>Median DFS is 38 months.</li>
</ul></li>
<li><strong>Treatment Distribution</strong>:
<ul>
<li>Radiotherapy (<code>Adj_Radio</code>) and chemotherapy (<code>Adj_Chem</code>) are evenly distributed across the population (~53% and ~48%, respectively).</li>
</ul></li>
</ul>
<div id="db29a50d-1063-4620-80be-0f0d0b947416" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Count missing values</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>display(cancer_df.isnull().<span class="bu">sum</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>Age (in years)     0
Dukes Stage        0
Gender             0
Location           0
DFS (in months)    0
DFS event          0
Adj_Radio          0
Adj_Chem           0
PC1                0
PC2                0
PC3                0
PC4                0
PC5                0
PC6                0
PC7                0
PC8                0
PC9                0
PC10               0
dtype: int64</code></pre>
</div>
</div>
<ul>
<li>Each column (<code>Age (in years)</code>, <code>Dukes Stage</code>, <code>Gender</code>, <code>Location</code>, <code>DFS (in months)</code>, <code>DFS event</code>, <code>Adj_Radio</code>, <code>Adj_Chem</code>, <code>PC1</code>-<code>PC10</code>) has <strong>0 missing entries</strong>.</li>
</ul>
<div id="7d5fe3b5-5b24-4c54-9aef-976697127333" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the histogram using Seaborn for better aesthetics</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>sns.histplot(</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>cancer_df, </span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span><span class="st">'DFS (in months)'</span>, </span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>    bins<span class="op">=</span><span class="dv">30</span>, </span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>    kde<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span><span class="st">'skyblue'</span>, </span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>    edgecolor<span class="op">=</span><span class="st">'black'</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Disease-Free Survival (DFS) Distribution"</span>, fontsize<span class="op">=</span><span class="dv">14</span>, fontweight<span class="op">=</span><span class="st">'bold'</span>)</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Disease-Free Survival (Months)"</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Frequency"</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>plt.xticks(fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>plt.yticks(fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a>plt.grid(axis<span class="op">=</span><span class="st">'y'</span>, linestyle<span class="op">=</span><span class="st">'--'</span>, alpha<span class="op">=</span><span class="fl">0.7</span>)</span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Colorectal-Cancer-Survival-Analysis_files/figure-html/cell-11-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<ul>
<li><strong>Distribution Shape</strong>:
<ul>
<li>The distribution of DFS (in months) is <strong>right-skewed</strong>, with most patients having shorter DFS durations.</li>
<li>A small number of patients have DFS extending to 100+ months.</li>
</ul></li>
<li><strong>Frequency Peaks</strong>:
<ul>
<li>There are noticeable peaks around 10–20 months and 50–60 months, indicating clusters of patients with shorter and moderate DFS durations.</li>
</ul></li>
<li><strong>Spread</strong>:
<ul>
<li>DFS durations range from <strong>4 months (minimum)</strong> to <strong>108 months (maximum)</strong>.</li>
<li>Most values are concentrated below 60 months, as indicated by the density curve.</li>
</ul></li>
</ul>
<div id="59ecb76a-85fc-4245-bd11-2db990448f37" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Count the values for DFS event</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>dfs_event_counts <span class="op">=</span> cancer_df[<span class="st">'DFS event'</span>].value_counts()</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">7</span>, <span class="dv">7</span>))</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>colors <span class="op">=</span> [<span class="st">'skyblue'</span>, <span class="st">'green'</span>]</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>explode <span class="op">=</span> [<span class="fl">0.05</span>, <span class="dv">0</span>]</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>ax.pie(</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>    dfs_event_counts,</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>    labels<span class="op">=</span>[<span class="st">'Censored (0)'</span>, <span class="st">'Event (1)'</span>],</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>    autopct<span class="op">=</span><span class="st">'</span><span class="sc">%1.1f%%</span><span class="st">'</span>,</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>    startangle<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>    explode<span class="op">=</span>explode,</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>    colors<span class="op">=</span>colors,</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>    shadow<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>    wedgeprops<span class="op">=</span>{<span class="st">'edgecolor'</span>: <span class="st">'black'</span>}</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>ax.set_title(<span class="st">"DFS Event Distribution"</span>, fontsize<span class="op">=</span><span class="dv">14</span>, fontweight<span class="op">=</span><span class="st">'bold'</span>)</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Colorectal-Cancer-Survival-Analysis_files/figure-html/cell-12-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<ul>
<li><strong>Distribution of Events</strong>:
<ul>
<li><strong>Censored (0)</strong>: Represents <strong>59.7%</strong> of the data, indicating these patients did not experience a DFS event (e.g., recurrence or death) during the observation period.</li>
<li><strong>Event (1)</strong>: Represents <strong>40.3%</strong> of the data, indicating these patients experienced a DFS event.</li>
</ul></li>
<li><strong>Balance</strong>:
<ul>
<li>The dataset shows a relatively balanced distribution between censored and event occurrences, with a slight majority in the censored group.</li>
</ul></li>
<li><strong>Implications for Analysis</strong>:
<ul>
<li>This balanced proportion provides a robust dataset for survival analysis, as both censored and event groups are adequately represented.</li>
</ul></li>
</ul>
</section>
<section id="preprocessing" class="level2">
<h2 class="anchored" data-anchor-id="preprocessing"><strong>Preprocessing</strong></h2>
<table class="caption-top table">
<colgroup>
<col style="width: 10%">
<col style="width: 30%">
<col style="width: 20%">
<col style="width: 37%">
</colgroup>
<thead>
<tr class="header">
<th><strong>Feature Type</strong></th>
<th><strong>Columns</strong></th>
<th><strong>Transformation</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Passthrough Features</strong></td>
<td><code>DFS (in months)</code>, <code>DFS event</code>, <code>Age (in years)</code>, <code>Adj_Radio</code>, <code>Adj_Chem</code></td>
<td>Passthrough</td>
<td>These features are kept as is, without any transformations.</td>
</tr>
<tr class="even">
<td><strong>Categorical Features</strong></td>
<td><code>Location</code></td>
<td>OneHotEncoder</td>
<td>Encodes categories into one-hot vectors. Handles unknown categories using <code>handle_unknown="ignore"</code>.</td>
</tr>
<tr class="odd">
<td><strong>Binary Features</strong></td>
<td><code>Gender</code></td>
<td>OneHotEncoder with <code>drop="if_binary"</code></td>
<td>Encodes binary features into one-hot vectors and drops one level to avoid redundancy.</td>
</tr>
<tr class="even">
<td><strong>Ordinal Features</strong></td>
<td><code>Dukes Stage</code></td>
<td>OrdinalEncoder</td>
<td>Encodes the ordered categories (<code>['A', 'B', 'C', 'D']</code>) into integers.</td>
</tr>
<tr class="odd">
<td><strong>Numeric Features</strong></td>
<td><code>PC1</code>, <code>PC2</code>, …, <code>PCn</code> (from Principal Component Analysis)</td>
<td>StandardScaler</td>
<td>Standardizes features to have a mean of 0 and a standard deviation of 1.</td>
</tr>
</tbody>
</table>
<div id="deae737f-298e-4a46-a2ac-5c4b347d5781" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define feature groups</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>passthrough_features <span class="op">=</span> [<span class="st">'DFS (in months)'</span>, <span class="st">'DFS event'</span>, <span class="st">'Age (in years)'</span>, <span class="st">'Adj_Radio'</span>, <span class="st">'Adj_Chem'</span>]</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>categorical_features <span class="op">=</span> [<span class="st">'Location'</span>]</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>binary_features <span class="op">=</span> [<span class="st">'Gender'</span>]</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>ordinary_features <span class="op">=</span> [<span class="st">'Dukes Stage'</span>]</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>numeric_features <span class="op">=</span> [<span class="ss">f'PC</span><span class="sc">{</span>i<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">'</span> <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(gene_data_pca.shape[<span class="dv">1</span>])]</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>preprocessor <span class="op">=</span> make_column_transformer(</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>    (</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>        <span class="st">"passthrough"</span>,</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>        passthrough_features,</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>    (</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>        OneHotEncoder(handle_unknown<span class="op">=</span><span class="st">"ignore"</span>),</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>        categorical_features,</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>    (</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>        OneHotEncoder(drop<span class="op">=</span><span class="st">"if_binary"</span>, handle_unknown<span class="op">=</span><span class="st">"ignore"</span>),</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>        binary_features,</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a>    (</span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a>        OrdinalEncoder(categories<span class="op">=</span>[[<span class="st">'A'</span>, <span class="st">'B'</span>, <span class="st">'C'</span>, <span class="st">'D'</span>]]),</span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a>        ordinary_features,</span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>    (</span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a>        StandardScaler(),</span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a>        numeric_features,</span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="07c25040-56c9-4472-ba00-3b4857e27af4" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>preprocessor.fit(cancer_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="13">
<style>#sk-container-id-1 {
  /* Definition of color scheme common for light and dark mode */
  --sklearn-color-text: black;
  --sklearn-color-line: gray;
  /* Definition of color scheme for unfitted estimators */
  --sklearn-color-unfitted-level-0: #fff5e6;
  --sklearn-color-unfitted-level-1: #f6e4d2;
  --sklearn-color-unfitted-level-2: #ffe0b3;
  --sklearn-color-unfitted-level-3: chocolate;
  /* Definition of color scheme for fitted estimators */
  --sklearn-color-fitted-level-0: #f0f8ff;
  --sklearn-color-fitted-level-1: #d4ebff;
  --sklearn-color-fitted-level-2: #b3dbfd;
  --sklearn-color-fitted-level-3: cornflowerblue;

  /* Specific color for light theme */
  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));
  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));
  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));
  --sklearn-color-icon: #696969;

  @media (prefers-color-scheme: dark) {
    /* Redefinition of color scheme for dark theme */
    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));
    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));
    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));
    --sklearn-color-icon: #878787;
  }
}

#sk-container-id-1 {
  color: var(--sklearn-color-text);
}

#sk-container-id-1 pre {
  padding: 0;
}

#sk-container-id-1 input.sk-hidden--visually {
  border: 0;
  clip: rect(1px 1px 1px 1px);
  clip: rect(1px, 1px, 1px, 1px);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

#sk-container-id-1 div.sk-dashed-wrapped {
  border: 1px dashed var(--sklearn-color-line);
  margin: 0 0.4em 0.5em 0.4em;
  box-sizing: border-box;
  padding-bottom: 0.4em;
  background-color: var(--sklearn-color-background);
}

#sk-container-id-1 div.sk-container {
  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`
     but bootstrap.min.css set `[hidden] { display: none !important; }`
     so we also need the `!important` here to be able to override the
     default hidden behavior on the sphinx rendered scikit-learn.org.
     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */
  display: inline-block !important;
  position: relative;
}

#sk-container-id-1 div.sk-text-repr-fallback {
  display: none;
}

div.sk-parallel-item,
div.sk-serial,
div.sk-item {
  /* draw centered vertical line to link estimators */
  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));
  background-size: 2px 100%;
  background-repeat: no-repeat;
  background-position: center center;
}

/* Parallel-specific style estimator block */

#sk-container-id-1 div.sk-parallel-item::after {
  content: "";
  width: 100%;
  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);
  flex-grow: 1;
}

#sk-container-id-1 div.sk-parallel {
  display: flex;
  align-items: stretch;
  justify-content: center;
  background-color: var(--sklearn-color-background);
  position: relative;
}

#sk-container-id-1 div.sk-parallel-item {
  display: flex;
  flex-direction: column;
}

#sk-container-id-1 div.sk-parallel-item:first-child::after {
  align-self: flex-end;
  width: 50%;
}

#sk-container-id-1 div.sk-parallel-item:last-child::after {
  align-self: flex-start;
  width: 50%;
}

#sk-container-id-1 div.sk-parallel-item:only-child::after {
  width: 0;
}

/* Serial-specific style estimator block */

#sk-container-id-1 div.sk-serial {
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: var(--sklearn-color-background);
  padding-right: 1em;
  padding-left: 1em;
}


/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is
clickable and can be expanded/collapsed.
- Pipeline and ColumnTransformer use this feature and define the default style
- Estimators will overwrite some part of the style using the `sk-estimator` class
*/

/* Pipeline and ColumnTransformer style (default) */

#sk-container-id-1 div.sk-toggleable {
  /* Default theme specific background. It is overwritten whether we have a
  specific estimator or a Pipeline/ColumnTransformer */
  background-color: var(--sklearn-color-background);
}

/* Toggleable label */
#sk-container-id-1 label.sk-toggleable__label {
  cursor: pointer;
  display: block;
  width: 100%;
  margin-bottom: 0;
  padding: 0.5em;
  box-sizing: border-box;
  text-align: center;
}

#sk-container-id-1 label.sk-toggleable__label-arrow:before {
  /* Arrow on the left of the label */
  content: "▸";
  float: left;
  margin-right: 0.25em;
  color: var(--sklearn-color-icon);
}

#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {
  color: var(--sklearn-color-text);
}

/* Toggleable content - dropdown */

#sk-container-id-1 div.sk-toggleable__content {
  max-height: 0;
  max-width: 0;
  overflow: hidden;
  text-align: left;
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-0);
}

#sk-container-id-1 div.sk-toggleable__content.fitted {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-0);
}

#sk-container-id-1 div.sk-toggleable__content pre {
  margin: 0.2em;
  border-radius: 0.25em;
  color: var(--sklearn-color-text);
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-0);
}

#sk-container-id-1 div.sk-toggleable__content.fitted pre {
  /* unfitted */
  background-color: var(--sklearn-color-fitted-level-0);
}

#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {
  /* Expand drop-down */
  max-height: 200px;
  max-width: 100%;
  overflow: auto;
}

#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {
  content: "▾";
}

/* Pipeline/ColumnTransformer-specific style */

#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {
  color: var(--sklearn-color-text);
  background-color: var(--sklearn-color-unfitted-level-2);
}

#sk-container-id-1 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {
  background-color: var(--sklearn-color-fitted-level-2);
}

/* Estimator-specific style */

/* Colorize estimator box */
#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-2);
}

#sk-container-id-1 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-2);
}

#sk-container-id-1 div.sk-label label.sk-toggleable__label,
#sk-container-id-1 div.sk-label label {
  /* The background is the default theme color */
  color: var(--sklearn-color-text-on-default-background);
}

/* On hover, darken the color of the background */
#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {
  color: var(--sklearn-color-text);
  background-color: var(--sklearn-color-unfitted-level-2);
}

/* Label box, darken color on hover, fitted */
#sk-container-id-1 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {
  color: var(--sklearn-color-text);
  background-color: var(--sklearn-color-fitted-level-2);
}

/* Estimator label */

#sk-container-id-1 div.sk-label label {
  font-family: monospace;
  font-weight: bold;
  display: inline-block;
  line-height: 1.2em;
}

#sk-container-id-1 div.sk-label-container {
  text-align: center;
}

/* Estimator-specific */
#sk-container-id-1 div.sk-estimator {
  font-family: monospace;
  border: 1px dotted var(--sklearn-color-border-box);
  border-radius: 0.25em;
  box-sizing: border-box;
  margin-bottom: 0.5em;
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-0);
}

#sk-container-id-1 div.sk-estimator.fitted {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-0);
}

/* on hover */
#sk-container-id-1 div.sk-estimator:hover {
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-2);
}

#sk-container-id-1 div.sk-estimator.fitted:hover {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-2);
}

/* Specification for estimator info (e.g. "i" and "?") */

/* Common style for "i" and "?" */

.sk-estimator-doc-link,
a:link.sk-estimator-doc-link,
a:visited.sk-estimator-doc-link {
  float: right;
  font-size: smaller;
  line-height: 1em;
  font-family: monospace;
  background-color: var(--sklearn-color-background);
  border-radius: 1em;
  height: 1em;
  width: 1em;
  text-decoration: none !important;
  margin-left: 1ex;
  /* unfitted */
  border: var(--sklearn-color-unfitted-level-1) 1pt solid;
  color: var(--sklearn-color-unfitted-level-1);
}

.sk-estimator-doc-link.fitted,
a:link.sk-estimator-doc-link.fitted,
a:visited.sk-estimator-doc-link.fitted {
  /* fitted */
  border: var(--sklearn-color-fitted-level-1) 1pt solid;
  color: var(--sklearn-color-fitted-level-1);
}

/* On hover */
div.sk-estimator:hover .sk-estimator-doc-link:hover,
.sk-estimator-doc-link:hover,
div.sk-label-container:hover .sk-estimator-doc-link:hover,
.sk-estimator-doc-link:hover {
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-3);
  color: var(--sklearn-color-background);
  text-decoration: none;
}

div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,
.sk-estimator-doc-link.fitted:hover,
div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,
.sk-estimator-doc-link.fitted:hover {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-3);
  color: var(--sklearn-color-background);
  text-decoration: none;
}

/* Span, style for the box shown on hovering the info icon */
.sk-estimator-doc-link span {
  display: none;
  z-index: 9999;
  position: relative;
  font-weight: normal;
  right: .2ex;
  padding: .5ex;
  margin: .5ex;
  width: min-content;
  min-width: 20ex;
  max-width: 50ex;
  color: var(--sklearn-color-text);
  box-shadow: 2pt 2pt 4pt #999;
  /* unfitted */
  background: var(--sklearn-color-unfitted-level-0);
  border: .5pt solid var(--sklearn-color-unfitted-level-3);
}

.sk-estimator-doc-link.fitted span {
  /* fitted */
  background: var(--sklearn-color-fitted-level-0);
  border: var(--sklearn-color-fitted-level-3);
}

.sk-estimator-doc-link:hover span {
  display: block;
}

/* "?"-specific style due to the `<a>` HTML tag */

#sk-container-id-1 a.estimator_doc_link {
  float: right;
  font-size: 1rem;
  line-height: 1em;
  font-family: monospace;
  background-color: var(--sklearn-color-background);
  border-radius: 1rem;
  height: 1rem;
  width: 1rem;
  text-decoration: none;
  /* unfitted */
  color: var(--sklearn-color-unfitted-level-1);
  border: var(--sklearn-color-unfitted-level-1) 1pt solid;
}

#sk-container-id-1 a.estimator_doc_link.fitted {
  /* fitted */
  border: var(--sklearn-color-fitted-level-1) 1pt solid;
  color: var(--sklearn-color-fitted-level-1);
}

/* On hover */
#sk-container-id-1 a.estimator_doc_link:hover {
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-3);
  color: var(--sklearn-color-background);
  text-decoration: none;
}

#sk-container-id-1 a.estimator_doc_link.fitted:hover {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-3);
}
</style><div id="sk-container-id-1" class="sk-top-container"><div class="sk-text-repr-fallback"><pre>ColumnTransformer(transformers=[('passthrough', 'passthrough',
                                 ['DFS (in months)', 'DFS event',
                                  'Age (in years)', 'Adj_Radio', 'Adj_Chem']),
                                ('onehotencoder-1',
                                 OneHotEncoder(handle_unknown='ignore'),
                                 ['Location']),
                                ('onehotencoder-2',
                                 OneHotEncoder(drop='if_binary',
                                               handle_unknown='ignore'),
                                 ['Gender']),
                                ('ordinalencoder',
                                 OrdinalEncoder(categories=[['A', 'B', 'C',
                                                             'D']]),
                                 ['Dukes Stage']),
                                ('standardscaler', StandardScaler(),
                                 ['PC1', 'PC2', 'PC3', 'PC4', 'PC5', 'PC6',
                                  'PC7', 'PC8', 'PC9', 'PC10'])])</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br>On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class="sk-container" hidden=""><div class="sk-item sk-dashed-wrapped"><div class="sk-label-container"><div class="sk-label fitted sk-toggleable"><input class="sk-toggleable__control sk-hidden--visually" id="sk-estimator-id-1" type="checkbox"><label for="sk-estimator-id-1" class="sk-toggleable__label fitted sk-toggleable__label-arrow fitted">&nbsp;&nbsp;ColumnTransformer<a class="sk-estimator-doc-link fitted" rel="noreferrer" target="_blank" href="https://scikit-learn.org/1.5/modules/generated/sklearn.compose.ColumnTransformer.html">?<span>Documentation for ColumnTransformer</span></a><span class="sk-estimator-doc-link fitted">i<span>Fitted</span></span></label><div class="sk-toggleable__content fitted"><pre>ColumnTransformer(transformers=[('passthrough', 'passthrough',
                                 ['DFS (in months)', 'DFS event',
                                  'Age (in years)', 'Adj_Radio', 'Adj_Chem']),
                                ('onehotencoder-1',
                                 OneHotEncoder(handle_unknown='ignore'),
                                 ['Location']),
                                ('onehotencoder-2',
                                 OneHotEncoder(drop='if_binary',
                                               handle_unknown='ignore'),
                                 ['Gender']),
                                ('ordinalencoder',
                                 OrdinalEncoder(categories=[['A', 'B', 'C',
                                                             'D']]),
                                 ['Dukes Stage']),
                                ('standardscaler', StandardScaler(),
                                 ['PC1', 'PC2', 'PC3', 'PC4', 'PC5', 'PC6',
                                  'PC7', 'PC8', 'PC9', 'PC10'])])</pre></div> </div></div><div class="sk-parallel"><div class="sk-parallel-item"><div class="sk-item"><div class="sk-label-container"><div class="sk-label fitted sk-toggleable"><input class="sk-toggleable__control sk-hidden--visually" id="sk-estimator-id-2" type="checkbox"><label for="sk-estimator-id-2" class="sk-toggleable__label fitted sk-toggleable__label-arrow fitted">passthrough</label><div class="sk-toggleable__content fitted"><pre>['DFS (in months)', 'DFS event', 'Age (in years)', 'Adj_Radio', 'Adj_Chem']</pre></div> </div></div><div class="sk-serial"><div class="sk-item"><div class="sk-estimator fitted sk-toggleable"><input class="sk-toggleable__control sk-hidden--visually" id="sk-estimator-id-3" type="checkbox"><label for="sk-estimator-id-3" class="sk-toggleable__label fitted sk-toggleable__label-arrow fitted">passthrough</label><div class="sk-toggleable__content fitted"><pre>passthrough</pre></div> </div></div></div></div></div><div class="sk-parallel-item"><div class="sk-item"><div class="sk-label-container"><div class="sk-label fitted sk-toggleable"><input class="sk-toggleable__control sk-hidden--visually" id="sk-estimator-id-4" type="checkbox"><label for="sk-estimator-id-4" class="sk-toggleable__label fitted sk-toggleable__label-arrow fitted">onehotencoder-1</label><div class="sk-toggleable__content fitted"><pre>['Location']</pre></div> </div></div><div class="sk-serial"><div class="sk-item"><div class="sk-estimator fitted sk-toggleable"><input class="sk-toggleable__control sk-hidden--visually" id="sk-estimator-id-5" type="checkbox"><label for="sk-estimator-id-5" class="sk-toggleable__label fitted sk-toggleable__label-arrow fitted">&nbsp;OneHotEncoder<a class="sk-estimator-doc-link fitted" rel="noreferrer" target="_blank" href="https://scikit-learn.org/1.5/modules/generated/sklearn.preprocessing.OneHotEncoder.html">?<span>Documentation for OneHotEncoder</span></a></label><div class="sk-toggleable__content fitted"><pre>OneHotEncoder(handle_unknown='ignore')</pre></div> </div></div></div></div></div><div class="sk-parallel-item"><div class="sk-item"><div class="sk-label-container"><div class="sk-label fitted sk-toggleable"><input class="sk-toggleable__control sk-hidden--visually" id="sk-estimator-id-6" type="checkbox"><label for="sk-estimator-id-6" class="sk-toggleable__label fitted sk-toggleable__label-arrow fitted">onehotencoder-2</label><div class="sk-toggleable__content fitted"><pre>['Gender']</pre></div> </div></div><div class="sk-serial"><div class="sk-item"><div class="sk-estimator fitted sk-toggleable"><input class="sk-toggleable__control sk-hidden--visually" id="sk-estimator-id-7" type="checkbox"><label for="sk-estimator-id-7" class="sk-toggleable__label fitted sk-toggleable__label-arrow fitted">&nbsp;OneHotEncoder<a class="sk-estimator-doc-link fitted" rel="noreferrer" target="_blank" href="https://scikit-learn.org/1.5/modules/generated/sklearn.preprocessing.OneHotEncoder.html">?<span>Documentation for OneHotEncoder</span></a></label><div class="sk-toggleable__content fitted"><pre>OneHotEncoder(drop='if_binary', handle_unknown='ignore')</pre></div> </div></div></div></div></div><div class="sk-parallel-item"><div class="sk-item"><div class="sk-label-container"><div class="sk-label fitted sk-toggleable"><input class="sk-toggleable__control sk-hidden--visually" id="sk-estimator-id-8" type="checkbox"><label for="sk-estimator-id-8" class="sk-toggleable__label fitted sk-toggleable__label-arrow fitted">ordinalencoder</label><div class="sk-toggleable__content fitted"><pre>['Dukes Stage']</pre></div> </div></div><div class="sk-serial"><div class="sk-item"><div class="sk-estimator fitted sk-toggleable"><input class="sk-toggleable__control sk-hidden--visually" id="sk-estimator-id-9" type="checkbox"><label for="sk-estimator-id-9" class="sk-toggleable__label fitted sk-toggleable__label-arrow fitted">&nbsp;OrdinalEncoder<a class="sk-estimator-doc-link fitted" rel="noreferrer" target="_blank" href="https://scikit-learn.org/1.5/modules/generated/sklearn.preprocessing.OrdinalEncoder.html">?<span>Documentation for OrdinalEncoder</span></a></label><div class="sk-toggleable__content fitted"><pre>OrdinalEncoder(categories=[['A', 'B', 'C', 'D']])</pre></div> </div></div></div></div></div><div class="sk-parallel-item"><div class="sk-item"><div class="sk-label-container"><div class="sk-label fitted sk-toggleable"><input class="sk-toggleable__control sk-hidden--visually" id="sk-estimator-id-10" type="checkbox"><label for="sk-estimator-id-10" class="sk-toggleable__label fitted sk-toggleable__label-arrow fitted">standardscaler</label><div class="sk-toggleable__content fitted"><pre>['PC1', 'PC2', 'PC3', 'PC4', 'PC5', 'PC6', 'PC7', 'PC8', 'PC9', 'PC10']</pre></div> </div></div><div class="sk-serial"><div class="sk-item"><div class="sk-estimator fitted sk-toggleable"><input class="sk-toggleable__control sk-hidden--visually" id="sk-estimator-id-11" type="checkbox"><label for="sk-estimator-id-11" class="sk-toggleable__label fitted sk-toggleable__label-arrow fitted">&nbsp;StandardScaler<a class="sk-estimator-doc-link fitted" rel="noreferrer" target="_blank" href="https://scikit-learn.org/1.5/modules/generated/sklearn.preprocessing.StandardScaler.html">?<span>Documentation for StandardScaler</span></a></label><div class="sk-toggleable__content fitted"><pre>StandardScaler()</pre></div> </div></div></div></div></div></div></div></div></div>
</div>
</div>
<div id="c1462cda-a115-4c16-8347-54df03072ef7" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>new_columns <span class="op">=</span> (</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>    passthrough_features</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>    <span class="op">+</span> preprocessor.named_transformers_[<span class="st">"onehotencoder-1"</span>]</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>    .get_feature_names_out(categorical_features)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>    .tolist()</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>    <span class="op">+</span> preprocessor.named_transformers_[<span class="st">"onehotencoder-2"</span>]</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>    .get_feature_names_out(binary_features)</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>    .tolist()</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>    <span class="op">+</span> preprocessor.named_transformers_[<span class="st">"ordinalencoder"</span>]</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>    .get_feature_names_out(ordinary_features)</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>    .tolist()</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>    <span class="op">+</span> numeric_features</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="b2f465ff-f861-4d8c-81cd-ab7b61c88389" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>cancer_df_surv <span class="op">=</span> pd.DataFrame(</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>    preprocessor.transform(cancer_df), index<span class="op">=</span>cancer_df.index, columns<span class="op">=</span>new_columns</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="b6ac63a7-c3e8-4590-aee1-093a6a864575" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>display(cancer_df_surv.head())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">DFS (in months)</th>
<th data-quarto-table-cell-role="th">DFS event</th>
<th data-quarto-table-cell-role="th">Age (in years)</th>
<th data-quarto-table-cell-role="th">Adj_Radio</th>
<th data-quarto-table-cell-role="th">Adj_Chem</th>
<th data-quarto-table-cell-role="th">Location_Colon</th>
<th data-quarto-table-cell-role="th">Location_Left</th>
<th data-quarto-table-cell-role="th">Location_Rectum</th>
<th data-quarto-table-cell-role="th">Location_Right</th>
<th data-quarto-table-cell-role="th">Gender_Male</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">PC1</th>
<th data-quarto-table-cell-role="th">PC2</th>
<th data-quarto-table-cell-role="th">PC3</th>
<th data-quarto-table-cell-role="th">PC4</th>
<th data-quarto-table-cell-role="th">PC5</th>
<th data-quarto-table-cell-role="th">PC6</th>
<th data-quarto-table-cell-role="th">PC7</th>
<th data-quarto-table-cell-role="th">PC8</th>
<th data-quarto-table-cell-role="th">PC9</th>
<th data-quarto-table-cell-role="th">PC10</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>108.0</td>
<td>0.0</td>
<td>62.0</td>
<td>1.0</td>
<td>0.0</td>
<td>0.0</td>
<td>1.0</td>
<td>0.0</td>
<td>0.0</td>
<td>1.0</td>
<td>...</td>
<td>1.690371</td>
<td>-0.474292</td>
<td>0.105510</td>
<td>1.029545</td>
<td>1.554643</td>
<td>0.700039</td>
<td>0.291275</td>
<td>-0.480326</td>
<td>-0.367550</td>
<td>0.175293</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>40.0</td>
<td>1.0</td>
<td>77.0</td>
<td>1.0</td>
<td>0.0</td>
<td>0.0</td>
<td>1.0</td>
<td>0.0</td>
<td>0.0</td>
<td>1.0</td>
<td>...</td>
<td>0.665861</td>
<td>1.091053</td>
<td>-1.810395</td>
<td>-2.216717</td>
<td>-0.398797</td>
<td>1.293620</td>
<td>0.710828</td>
<td>1.273473</td>
<td>0.417849</td>
<td>1.620073</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>49.0</td>
<td>0.0</td>
<td>66.0</td>
<td>1.0</td>
<td>0.0</td>
<td>0.0</td>
<td>1.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>...</td>
<td>0.972914</td>
<td>-0.611976</td>
<td>-0.288582</td>
<td>0.450546</td>
<td>0.078611</td>
<td>0.154671</td>
<td>-0.304219</td>
<td>0.886651</td>
<td>-2.502641</td>
<td>0.852376</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>45.0</td>
<td>0.0</td>
<td>72.0</td>
<td>1.0</td>
<td>1.0</td>
<td>0.0</td>
<td>1.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>...</td>
<td>2.033161</td>
<td>-0.661475</td>
<td>0.031354</td>
<td>0.237474</td>
<td>-0.086228</td>
<td>-0.423905</td>
<td>-0.435577</td>
<td>1.715649</td>
<td>0.278953</td>
<td>-1.615547</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>40.0</td>
<td>0.0</td>
<td>75.0</td>
<td>0.0</td>
<td>1.0</td>
<td>0.0</td>
<td>1.0</td>
<td>0.0</td>
<td>0.0</td>
<td>1.0</td>
<td>...</td>
<td>0.118596</td>
<td>1.097989</td>
<td>0.853899</td>
<td>-0.444289</td>
<td>0.760363</td>
<td>0.017561</td>
<td>1.353270</td>
<td>0.819576</td>
<td>0.115883</td>
<td>0.048644</td>
</tr>
</tbody>
</table>

<p>5 rows × 21 columns</p>
</div>
</div>
</div>
</section>
<section id="kaplan-meier-survival-curve" class="level2">
<h2 class="anchored" data-anchor-id="kaplan-meier-survival-curve"><strong>Kaplan-Meier survival curve</strong></h2>
<div id="e26d41b9-af53-4514-bc78-3d2649c30999" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>kmf <span class="op">=</span> lifelines.KaplanMeierFitter()</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>kmf.fit(cancer_df_surv[<span class="st">"DFS (in months)"</span>], cancer_df_surv[<span class="st">"DFS event"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="17">
<pre><code>&lt;lifelines.KaplanMeierFitter:"KM_estimate", fitted with 62 total observations, 25 right-censored observations&gt;</code></pre>
</div>
</div>
<section id="survival-curve" class="level3">
<h3 class="anchored" data-anchor-id="survival-curve"><strong>Survival curve</strong></h3>
<div id="dca2c192-2bfa-4d72-b250-ca6823bc9b95" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>kmf.survival_function_.plot()<span class="op">;</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Survival fuction of Disease-Free Survival (DFS)"</span>, fontsize<span class="op">=</span><span class="dv">14</span>, fontweight<span class="op">=</span><span class="st">'bold'</span>)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Time with survive (months)"</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Survival Probability"</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>plt.grid(<span class="va">True</span>, linestyle<span class="op">=</span><span class="st">'--'</span>, alpha<span class="op">=</span><span class="fl">0.7</span>)</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Colorectal-Cancer-Survival-Analysis_files/figure-html/cell-19-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>The Kaplan-Meier curve shows a <strong>gradual decline</strong> in survival probability, with the steepest drop in the first <strong>40 months</strong>.</li>
<li>The <strong>median DFS</strong> is approximately <strong>40 months</strong>, indicating that <strong>50% of patients</strong> experience disease recurrence or death by this time.</li>
<li>A small subset of patients maintains disease-free survival for <strong>100+ months</strong>, with survival probability stabilizing around <strong>20%</strong>.</li>
</ul>
</section>
<section id="error-bars" class="level3">
<h3 class="anchored" data-anchor-id="error-bars"><strong>Error bars</strong></h3>
<div id="ecdc5a2e-5169-41d7-b916-055035cde0d9" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>kmf.plot_survival_function()</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Survival fuction of Disease-Free Survival (DFS)"</span>, fontsize<span class="op">=</span><span class="dv">14</span>, fontweight<span class="op">=</span><span class="st">'bold'</span>)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Time with survive (months)"</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Survival Probability"</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>plt.grid(<span class="va">True</span>, linestyle<span class="op">=</span><span class="st">'--'</span>, alpha<span class="op">=</span><span class="fl">0.7</span>)</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Colorectal-Cancer-Survival-Analysis_files/figure-html/cell-20-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>At earlier time points (e.g., <strong>0–40 months</strong>), confidence intervals (shaded regions) are narrower, reflecting higher certainty due to a larger number of patients and observations.</li>
<li>As survival time increases (e.g., <strong>beyond 60 months</strong>), fewer patients remain under observation, resulting in wider intervals and greater uncertainty in survival probability estimates.</li>
</ul>
</section>
</section>
<section id="cox-proportional-hazards-model" class="level2">
<h2 class="anchored" data-anchor-id="cox-proportional-hazards-model"><strong>Cox proportional hazards model</strong></h2>
<div id="4e56b67a-2b29-4af8-a9ce-27502ad338fe" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>cph <span class="op">=</span> lifelines.CoxPHFitter(penalizer<span class="op">=</span><span class="fl">0.1</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>cph.fit(cancer_df_surv, duration_col<span class="op">=</span><span class="st">"DFS (in months)"</span>, event_col<span class="op">=</span><span class="st">"DFS event"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="20">
<pre><code>&lt;lifelines.CoxPHFitter: fitted with 62 total observations, 25 right-censored observations&gt;</code></pre>
</div>
</div>
<div id="707762a7-e5af-4d6c-95db-f2b50cad83ff" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">12</span>))</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>cph.plot()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Colorectal-Cancer-Survival-Analysis_files/figure-html/cell-22-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<ul>
<li><strong>Significant Predictors</strong>:
<ul>
<li><strong>Dukes Stage</strong>: Positive coefficient with a narrow CI that does not cross 0 too much, indicating a significant increase in hazard (increase the risk of recurrence or death).</li>
<li><strong>Adj_Radio</strong>: Narrow CI that does not cross 0 too much, suggesting a significant protective effect in reducing the hazard (reduce the risk of recurrence or death).</li>
<li>Many <strong>PC</strong> which represnet specifc gene types have either positive or negative coefficients are significant which suggests that genes are significant predictors.</li>
</ul></li>
<li><strong>Non-Significant Predictors</strong>:
<ul>
<li>Predictors like <strong>Gender_Male</strong>, <strong>Location_Colon</strong>, and <strong>PC9</strong> have confidence intervals that cross 0, indicating no statistically significant impact on hazard.</li>
</ul></li>
</ul>
<div id="bd6b17bf-a03a-4e90-9598-6a655c645b88" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>cph.plot_partial_effects_on_outcome(<span class="st">"Gender_Male"</span>, [<span class="dv">0</span>, <span class="dv">1</span>])</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Time with survive (months)"</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Survival Probability"</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>plt.gcf().set_size_inches(<span class="dv">10</span>, <span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Colorectal-Cancer-Survival-Analysis_files/figure-html/cell-23-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<ul>
<li><strong>Gender Differences in Survival</strong>:
<ul>
<li>The survival probability for <strong>males</strong> is slightly higher than for <strong>females</strong> across most of the observation period.</li>
<li>This indicates that males may have a marginally better survival rate in this dataset.</li>
</ul></li>
<li><strong>No Significant Divergence</strong>:
<ul>
<li>The survival curves for both genders follow a similar trend, with no substantial divergence observed, suggesting gender may not have a strong impact on survival.</li>
</ul></li>
</ul>
<div id="5bd07888-3813-4ab8-bdef-8022645ea488" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>cph.plot_partial_effects_on_outcome(<span class="st">"Dukes Stage"</span>, [<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>])</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Time with survive (months)"</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Survival Probability"</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>plt.gcf().set_size_inches(<span class="dv">10</span>, <span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Colorectal-Cancer-Survival-Analysis_files/figure-html/cell-24-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<ul>
<li><strong>Trend Over Time</strong>:
<ul>
<li>Survival probability decreases for all stages over time, but the rate of decline is faster for higher stages (e.g., Dukes Stage 2 and 3).</li>
<li>The survival curves are well-separated by stage, reflecting a clear impact of disease severity on survival outcomes.</li>
</ul></li>
<li><strong>Long-Term Survival</strong>:
<ul>
<li><strong>Dukes Stage 0 and 1</strong> stabilize at higher survival probabilities (~60-70%) after 80 months.</li>
<li><strong>Dukes Stage 3</strong> has the lowest long-term survival (~20%).</li>
</ul></li>
</ul>
<div id="da3bd032-30c1-4282-89a5-a72b70970336" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>cph.plot_partial_effects_on_outcome(<span class="st">'Adj_Radio'</span>, [<span class="dv">0</span>, <span class="dv">1</span>])</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Time with survive (months)"</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Survival Probability"</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>plt.gcf().set_size_inches(<span class="dv">10</span>, <span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Colorectal-Cancer-Survival-Analysis_files/figure-html/cell-25-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<ul>
<li><strong>Survival Probability with Radiotherapy</strong>:
<ul>
<li>Patients who <strong>received radiotherapy</strong> have consistently higher survival probabilities compared to those who <strong>did not receive radiotherapy</strong>.</li>
</ul></li>
<li><strong>Long-Term Survival</strong>:
<ul>
<li>Patients who <strong>received radiotherapy</strong> stabilize at a higher survival probability (~40%) after 80 months, while those w who <strong>did not receive radiotherapy</strong> have a survival probability below 20%, indicating that radiotherapy improves disease-free survival (<strong>DFS</strong>) time.</li>
</ul></li>
</ul>
<div id="3ee3d2c9-01d1-4e20-86f7-bc5526d45371" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>cph.plot_partial_effects_on_outcome(<span class="st">'Adj_Chem'</span>, [<span class="dv">0</span>, <span class="dv">1</span>])</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Time with survive (months)"</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Survival Probability"</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>plt.gcf().set_size_inches(<span class="dv">10</span>, <span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Colorectal-Cancer-Survival-Analysis_files/figure-html/cell-26-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<ul>
<li><strong>Survival Probability with Chemotherapy</strong>:
<ul>
<li>Patients who <strong>received chemotherapy</strong> have slightly lower survival probabilities compared to those who <strong>did not receive chemotherapy</strong>.</li>
</ul></li>
<li><strong>Long-Term Survival</strong>:
<ul>
<li>Both groups stabilize at similar survival probabilities (~20%) after 100 months, indicating minimal difference in long-term survival outcomes, suggesting that chemotherapy does not significantly impact disease-free survival (<strong>DFS</strong>) time.</li>
</ul></li>
</ul>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>